### PR TITLE
update syn + quote

### PR DESCRIPTION
**Pull request by nerdrew**
I'm not sure on lots of this, but the tests pass. I had multiple versions of syn in one of my projects because of diesel. Figured I'd just update it. It was more involved than I expected.

### CODE REVIEW

File: diesel_codegen/src/util.rs [hunk 1]
```diff
@@ -101,22 +101,29 @@ pub fn inner_of_option_ty(ty: &Ty) -> Option<&Ty> {
...
     match options {
-        Some(&MetaItem::List(_, ref options)) => Some(options),
+        Some(&MetaItem::List(_, ref options)) => {
+            Some(options.iter().map(|o| {
+                match o {
+                   &NestedMetaItem::MetaItem(ref m) => m.clone(),
+                   _ => on_bug(),
+                }
+            }).collect())
+        }
```
Comment (killercup) [inline diesel_codegen/src/util.rs]: This looks pretty complicated, cf. <https://github.com/diesel-rs/diesel/pull/508/files#r90534159>

File: diesel_tests/tests/expressions/ops.rs [hunk 2]
```diff
@@ -91,6 +91,6 @@ fn mix_and_match_all_numeric_ops() {
         (3, 'Jim'), (4, 'Bob')").unwrap();
 
     let expected_data = vec![4, 6, 7, 9];
-    let data = users.select(id * 3 / 2 + 4 - 1).load(&connection);
+    let data = users.select(id * 3 / 2 + 4 - 1).order(id).load(&connection);
```
Comment (sgrif) [inline diesel_tests/tests/expressions/ops.rs]: We have a bunch of intermittent failures from needing an order added, would definitely prefer in a separate PR. It affects most of our tests. Actually making me contemplate default ordering.

File: diesel_codegen/src/schema_inference.rs [hunk 3]
```diff
@@ -35,20 +35,20 @@ pub fn derive_infer_table_from_schema(input: syn::MacroInput) -> quote::Tokens {
...
-    let table_name = get_option(options, "table_name", bug);
+    let database_url = get_option(&options, "database_url", bug);
+    let table_name = get_option(&options, "table_name", bug);
 
-    let connection = establish_connection(database_url).unwrap();
-    let data = get_table_data(&connection, table_name).unwrap();
-    let primary_keys = get_primary_keys(&connection, table_name).unwrap()
+    let connection = establish_connection(database_url.as_str()).unwrap();
+    let data = get_table_data(&connection, table_name.as_str()).unwrap();
+    let primary_keys = get_primary_keys(&connection, table_name.as_str()).unwrap()
```
Comment (nerdrew) [inline diesel_codegen/src/schema_inference.rs]: No reason. What would you prefer?

File: diesel_tests/tests/types_roundtrip.rs [hunk 4]
```diff
@@ -138,12 +138,12 @@ pub fn mk_naive_date(days: u32) -> NaiveDate {
 #[cfg(feature = "postgres")]
 mod unstable_types {
     use super::*;
-    use std::time::*;
+    use std::time::{SystemTime, UNIX_EPOCH};
```
Comment (nerdrew) [inline diesel_tests/tests/types_roundtrip.rs]: `std::time::Duration` conflicts with `super::Duration`.

File: diesel_tests/tests/expressions/ops.rs (hunk 2 above)
Comment (nerdrew) [inline diesel_tests/tests/expressions/ops.rs]: Should I break out the ordering changes to the tests into a separate PR or do you want to fix it more holistically?

Comment (sgrif) [inline diesel_tests/tests/expressions/ops.rs]: If you want to break them out into a separate PR, that's fine. I want to evaluate the issue a bit more before I would merge it though. Either way it should be left out of this PR

File: diesel_codegen/src/util.rs [hunk 5]
```diff
@@ -101,22 +101,29 @@ pub fn inner_of_option_ty(ty: &Ty) -> Option<&Ty> {
...
         Some(_) => on_bug(),
         None => None,
     }
 }
 
-pub fn get_option<'a>(
-    options: &'a [MetaItem],
+pub fn get_option(
+    options: &Vec<MetaItem>,
```
Comment (nerdrew) [inline diesel_codegen/src/util.rs]: It isn't. I just couldn't get it to compile using references :(

File: diesel_tests/tests/expressions/ops.rs (hunk 2 above)
Comment (sgrif) [inline diesel_tests/tests/expressions/ops.rs]: As a general rule of thumb we try to keep things small, focused, and easy to review. For example, #508, #509, and #510 could have reasonably been a single PR, but it helps the reviewer a ton to keep things more granular.

File: diesel_codegen/src/util.rs (hunk 1 above)
Comment (nerdrew) [inline diesel_codegen/src/util.rs]: I think this is the change that prompted the `&str` to `String` and `&[..]` to `Vec` changes.

`MetaItem::List` needs to be transformed into a list of `MetaItem`'s. I couldn't figure out how to return a slice of these. Ideas?

File: diesel_codegen/src/schema_inference.rs (hunk 3 above)
Comment (sgrif) [inline diesel_codegen/src/schema_inference.rs]: `&x` is preferred to `x.as_str()`

File: diesel_codegen/src/util.rs (hunk 1 above)
Comment (sgrif) [inline diesel_codegen/src/util.rs]: This still shouldn't need to affect the input type.

File: diesel_codegen/src/util.rs (hunk 5 above)
Comment (sgrif) [inline diesel_codegen/src/util.rs]: I could see the return types needing to change. Unclear on why the input type should have needed to change. There is no case where `&Vec<MetaItem>` wouldn't deref to `&[MetaItem]`

File: diesel_codegen/src/embed_migrations.rs [hunk 6]
```diff
@@ -16,7 +16,7 @@ pub fn derive_embed_migrations(input: syn::MacroInput) -> quote::Tokens {
 
     let options = get_options_from_input(&input.attrs, bug);
     let migrations_path_opt = options.map(|o| get_option(&o, "migrations_path", bug));
```
Comment (nerdrew) [inline diesel_codegen/src/embed_migrations.rs]: Is there a way to pass the lifetime of `options` to `get_option` and use that as the lifetime on the returned `str`? I couldn't get that to work, which is why I switch the return value of `get_option` to `String` (and all the associated changes `str` to `String`).

**Comment by nerdrew** [comment]
General workflow question: do you prefer "fixup" commits and then squash before merging, or should I just amend this commit?

File: diesel_codegen/src/schema_inference.rs (hunk 3 above)
Comment (nerdrew) [inline diesel_codegen/src/schema_inference.rs]: Will fix.

**Comment by sgrif** [comment]
Separate commits during code review. I will often request a squash after everything looks good

File: diesel_codegen/src/util.rs (hunk 1 above)
Comment (nerdrew) [inline diesel_codegen/src/util.rs]: Yup. Fixed.

File: diesel_tests/tests/expressions/ops.rs (hunk 2 above)
Comment (nerdrew) [inline diesel_tests/tests/expressions/ops.rs]: Will remove the `order` changes. The rest seems to be atomic though, yes?

File: diesel_codegen/src/embed_migrations.rs [hunk 7]
```diff
@@ -16,7 +16,7 @@ pub fn derive_embed_migrations(input: syn::MacroInput) -> quote::Tokens {
...
-    let migrations_expr = migration_directory_from_given_path(migrations_path_opt)
+    let migrations_expr = migration_directory_from_given_path(migrations_path_opt.as_ref().map(|s| s.as_str()))
```
Comment (nerdrew) [inline diesel_codegen/src/embed_migrations.rs]: Is there a cleaner way to go from `Option<String>` to `Option<&str>`?

File: diesel_tests/tests/expressions/ops.rs (hunk 2 above)
Comment (sgrif) [inline diesel_tests/tests/expressions/ops.rs]: Yes, I think so

**Comment by sgrif** [comment]
Thanks for working on this. I have to go make dinner, but I will finish reviewing tomorrow

File: diesel_codegen/src/associations.rs [hunk 8]
```diff
@@ -75,10 +75,17 @@ fn build_association_options(
...
                 _ => return usage_error,
             };
-            let foreign_key_name = options.iter().find(|a| a.name() == "foreign_key")
+            let foreign_key_name = options.iter()
+                .filter_map(|o| {
+                    match o {
+                        &syn::NestedMetaItem::MetaItem(ref mi) => Some(mi),
+                        _ => None
+                    }
+                })
```
Comment (killercup) [inline diesel_codegen/src/associations.rs]: You can make this a bit more concise by the way:

```rust
.filter_map(|o| match o {
    &syn::NestedMetaItem::MetaItem(ref mi) => Some(mi),
    _ => None
})
```

File: diesel_codegen/src/embed_migrations.rs (hunk 7 above)
Comment (killercup) [inline diesel_codegen/src/embed_migrations.rs]: Not that I know of. I'd probably replace `.map(|s| s.as_str())` with `.map(String::as_ref)` though.

Edit: There even is a [SO question](http://stackoverflow.com/q/31233938/1254484) about that with basically the same answer.

File: diesel_codegen/src/embed_migrations.rs (hunk 6 above)
Comment (killercup) [inline diesel_codegen/src/embed_migrations.rs]: I think the problem is that `.map` creates a reference that is only valid in its closure, but you want to return a value that lives as long as `option` does (and not the reference to option). Try this:

```diff
- let migrations_path_opt = options.map(|o| get_option(&o, "migrations_path", bug));
+ let migrations_path_opt = options.as_ref().map(|o| get_option(o, "migrations_path", bug));
```

File: diesel_codegen/src/associations.rs (hunk 8 above)
Comment (nerdrew) [inline diesel_codegen/src/associations.rs]: Done

File: diesel_codegen/src/embed_migrations.rs (hunk 6 above)
Comment (nerdrew) [inline diesel_codegen/src/embed_migrations.rs]: Done.

File: diesel_codegen/src/embed_migrations.rs (hunk 7 above)
Comment (nerdrew) [inline diesel_codegen/src/embed_migrations.rs]: Good to know; unnecessary in this PR now though ;)

File: diesel_codegen/src/schema_inference.rs [hunk 9]
```diff
@@ -35,20 +35,20 @@ pub fn derive_infer_table_from_schema(input: syn::MacroInput) -> quote::Tokens {
...
-    let table_name = get_option(options, "table_name", bug);
+    let database_url = get_option(&options, "database_url", bug);
+    let table_name = get_option(&options, "table_name", bug);
 
-    let connection = establish_connection(database_url).unwrap();
-    let data = get_table_data(&connection, table_name).unwrap();
-    let primary_keys = get_primary_keys(&connection, table_name).unwrap()
+    let connection = establish_connection(&database_url).unwrap();
+    let data = get_table_data(&connection, &table_name).unwrap();
+    let primary_keys = get_primary_keys(&connection, &table_name).unwrap()
```
Comment (killercup) [inline diesel_codegen/src/schema_inference.rs]: Just for a fun exercise: I'm pretty sure the `&`s you added here are no longer necessary – the bindings are already references. (Deref coercion allows rustc to accept `&&&foo` where `&foo` was wanted.)

File: diesel_codegen/src/embed_migrations.rs [hunk 10]
```diff
@@ -15,7 +15,7 @@ pub fn derive_embed_migrations(input: syn::MacroInput) -> quote::Tokens {
     }
 
     let options = get_options_from_input(&input.attrs, bug);
-    let migrations_path_opt = options.map(|o| get_option(&o, "migrations_path", bug));
+    let migrations_path_opt = options.as_ref().map(|o| get_option(&o, "migrations_path", bug));
```
Comment (killercup) [inline diesel_codegen/src/embed_migrations.rs]: Nice. I think you can just give it `o` instead of `&o` here (same as other comment).

File: diesel_codegen/src/ast_builder.rs [hunk 11]
```diff
@@ -11,6 +11,6 @@ pub fn ty_path(path: Path) -> Ty {
 pub fn path_ident(ident: Ident) -> Path {
     Path {
         global: false,
-        segments: vec![PathSegment::ident(ident)],
+        segments: vec![PathSegment::from(ident)],
     }
```
Comment (killercup) [inline diesel_codegen/src/ast_builder.rs]: FYI this whole function can be replaced with `ident.into()`

File: diesel_codegen/src/schema_inference.rs [hunk 12]
```diff
@@ -62,7 +62,7 @@ fn column_def_tokens(
     let path_segments = column_type.path
         .into_iter()
         .map(syn::Ident::new)
-        .map(syn::PathSegment::ident)
+        .map(syn::PathSegment::from)
```
Comment (killercup) [inline diesel_codegen/src/schema_inference.rs]: I added a bunch of `From` impls to syn a while back. I don't think we need the `.map(syn::Ident::new)` here anymore.

**Comment by sgrif** [comment]
@killercup feel free to merge if you think it's good to go

**Comment by killercup** [comment]
@nerdrew, can you also have a look at the conflict and the CI failure? If you don't have time, that's okay, I can do it tomorrow.

Travis says:

```
error: unused variable: `lifetimes`
  --> /home/travis/build/diesel-rs/diesel/diesel_codegen/src/identifiable.rs:10:9
   |
10 |     let lifetimes = model.generics.lifetimes;
   |         ^^^^^^^^^
   |
```

which surprises me because

0. it's not a bug in cargo/rustc (yay, CI works again!),
1. this line is not in `identifiable.rs:10` but in `insertable.rs:21` (maybe proc macros messing up the source of the error),
2. `lifetimes` is passed to the `_Insertable!` as `(#(#lifetimes),*)`,
3. and it compiles fine locally with my nightly from 2016-09-28.

You can find the code in its current form [here](https://github.com/diesel-rs/diesel/blob/9d90d6d70dd1dc295231963854a0593aa4f95d8b/diesel_codegen/src/insertable.rs#L21-L29). Maybe adding `#[allow(unused_assignments)]` (or similar, with a comment) as a temporary solution is fine.

**Comment by sgrif** [comment]
I think the reason it's unused is that this branch was made before we had lifetimes there, but the file merged fine, so the "usage" is using the wrong syntax and therefore not actually being used.

File: diesel_codegen/src/schema_inference.rs (hunk 9 above)
Comment (nerdrew) [inline diesel_codegen/src/schema_inference.rs]: Just the `&`s for the `&str`s right? I still need them for the `options: Option<Vec<MetaItem>>` returned from `get_options_from_input`?

File: diesel_codegen/src/schema_inference.rs (hunk 12 above)
Comment (nerdrew) [inline diesel_codegen/src/schema_inference.rs]: Removed.

File: diesel_codegen/src/schema_inference.rs (hunk 9 above)
Comment (nerdrew) [inline diesel_codegen/src/schema_inference.rs]: TIL.

File: diesel_codegen/src/embed_migrations.rs (hunk 10 above)
Comment (nerdrew) [inline diesel_codegen/src/embed_migrations.rs]: Done.

File: diesel_codegen/src/ast_builder.rs (hunk 11 above)
Comment (nerdrew) [inline diesel_codegen/src/ast_builder.rs]: The whole `path_ident` function? I.e. `path_ident(ident)` => `ident.into()`?

Comment (nerdrew) [inline diesel_codegen/src/ast_builder.rs]: The method is public, does that matter in this case?

**Comment by nerdrew** [comment]
@sgrif ya, those "used variable" errors are caused by `quote` changes. The errors are real. Fixing now.

**Comment by sgrif** [comment]
Thanks!
